internal/stateindex/encoding.go

```diff
@@ -0,0 +1,96 @@
+package stateindex
+
+import (
+	"encoding/binary"
+	"fmt"
+	"math"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/pkg/errors"
+)
+
+// encodeOrderPreservingVarUint64 returns a byte-representation for a uint64 number such that
+// all zero-bits starting bytes are trimmed in order to reduce the length of the array
+// For preserving the order in a default bytes-comparison, first byte contains the number of remaining bytes.
+// The presence of first byte also allows to use the returned bytes as part of other larger byte array such as a
+// composite-key representation in db
+func encodeOrderPreservingVarUint64(number uint64) []byte {
+	bytes := make([]byte, 8)
+	binary.BigEndian.PutUint64(bytes, number)
+	startingIndex := 0
+	size := 0
+	for i, b := range bytes {
+		if b != 0x00 {
+			startingIndex = i
+			size = 8 - i
+			break
+		}
+	}
+	sizeBytes := proto.EncodeVarint(uint64(size))
+	if len(sizeBytes) > 1 {
+		panic(fmt.Errorf("[]sizeBytes should not be more than one byte because the max number it needs to hold is 8. size=%d", size))
+	}
+	encodedBytes := make([]byte, size+1)
+	encodedBytes[0] = sizeBytes[0]
+	copy(encodedBytes[1:], bytes[startingIndex:])
+	return encodedBytes
+}
```

Comment:
I think there is a faster way to do this.... instead of triple allocation and double copy:
-  implement the loop over the uint64 bytes as they do inside the `binary.BigEndian.PutUint64(bytes, number)` function, (shifting the value by multiples of 8 every round) 
- when you find the first non-zero byte allocate the right size of slice (size+1), 
- the varint encoding of the uint64 size is just byte(size)
- populate the first two bytes, and continue the loop writing one byte at a time

Reply:
Done.